spring-kafka/src/main/java/org/springframework/kafka/requestreply/AggregatingReplyingKafkaTemplate.java

```diff
@@ -0,0 +1,281 @@
+ *
+ * @author Gary Russell
+ * @since 2.3
+ *
+ */
+public class AggregatingReplyingKafkaTemplate<K, V, R>
+		extends ReplyingKafkaTemplate<K, V, Collection<ConsumerRecord<K, R>>>
+		implements BatchConsumerAwareMessageListener<K, Collection<ConsumerRecord<K, R>>> {
+
+	/**
+	 * Pseudo topic name for the "outer" {@link ConsumerRecords} that has the aggregated
+	 * results in its value after a normal release by the release strategy.
+	 */
+	public static final String AGGREGATED_RESULTS_TOPIC = "aggregatedResults";
+
+	/**
+	 * Pseudo topic name for the "outer" {@link ConsumerRecords} that has the aggregated
+	 * results in its value after a timeout.
+	 */
+	public static final String PARTIAL_RESULTS_AFTER_TIMEOUT_TOPIC = "partialResultsAfterTimeout";
+
+	private final ConcurrentMap<CorrelationKey, Set<RecordHolder<K, R>>> pending = new ConcurrentHashMap<>();
+
+	private final Map<TopicPartition, Long> offsets = new HashMap<>();
+
+	private final Predicate<Collection<ConsumerRecord<K, R>>> releaseStrategy;
+
+	private Duration commitTimeout = Duration.ofSeconds(30);
+
+	private boolean returnPartialOnTimeout;
+
+	private volatile long lastOrphanCheck = System.currentTimeMillis();
+
+	/**
+	 * Construct an instance using the provided parameter arguments. The releaseStrategy
+	 * is consulted to determine when a collection is "complete".
+	 * @param producerFactory the producer factory.
+	 * @param replyContainer the reply container.
+	 * @param releaseStrategy the release strategy.
+	 */
+	public AggregatingReplyingKafkaTemplate(ProducerFactory<K, V> producerFactory,
+			GenericMessageListenerContainer<K, Collection<ConsumerRecord<K, R>>> replyContainer,
+			Predicate<Collection<ConsumerRecord<K, R>>> releaseStrategy) {
+
+		super(producerFactory, replyContainer);
+		Assert.notNull(releaseStrategy, "'releaseStrategy' cannot be null");
+		AckMode ackMode = replyContainer.getContainerProperties().getAckMode();
+		Assert.isTrue(ackMode.equals(AckMode.MANUAL) || ackMode.equals(AckMode.MANUAL_IMMEDIATE),
+				"The reply container must have a MANUAL or MANUAL_IMMEDIATE AckMode");
+		this.releaseStrategy = releaseStrategy;
+	}
+
+	/**
+	 * Set the timeout to use when committing offsets.
+	 * @param commitTimeout the timeout.
+	 */
+	public void setCommitTimeout(Duration commitTimeout) {
+		Assert.notNull(commitTimeout, "'commitTimeout' cannot be null");
+		this.commitTimeout = commitTimeout;
+	}
+
+	/**
+	 * Set to true to return a partial result when a request times out.
+	 * @param returnPartialOnTimeout true to return a partial result.
+	 */
+	public void setReturnPartialOnTimeout(boolean returnPartialOnTimeout) {
+		this.returnPartialOnTimeout = returnPartialOnTimeout;
+	}
+
+	@Override
+	public void onMessage(List<ConsumerRecord<K, Collection<ConsumerRecord<K, R>>>> data, Consumer<?, ?> consumer) {
+		long now = System.currentTimeMillis();
+		if (now - this.lastOrphanCheck > getReplyTimeout() * 10) { // NOSONAR magic #
+			weedOrphans(consumer);
```

Comment:
What would that add?

If we get a real timeout, `handleTimeout()` removes the orphaned replies for that specific request (optionally returning a partial result).

Reply:
Oh! Right. So, then the opposite story question: why do we need to check for orphans in the `onMessage` then?

Won't that `handleTimeout` for all the orphans cover us?

Reply:
No; consider this case - request with 2 replies...

- request1
- request2
- request3
- reply1a
- reply2a
- reply3a
- reply1b (complete group)
- request2 times out (and clears out reply2a)
- rebalance

Since we had a pending request3, we didn't commit the offsets for the replies to request1.

After the rebalance, we get

- reply1a
- reply2a
- reply3a
- reply1b (complete group)
- reply3b (complete group)

request1 replies will be passed to super.onMessage, but that's ok, it will log 

```java
"No pending reply: " + record + WITH_CORRELATION_ID
				+ correlationId + ", perhaps timed out, or using a shared reply topic";
```

reply3b will complete request3.

But the reply2a redelivery will stay forever because request2 already timed out so `handleTimeout()` won't be called again.

Reply:
Huh? Why should we worry about `reply2a` then if there is no `request2` for it any more?

We should log "late reply" and discard. What else am I missing?

Your explanation is good, but I'm not sure what is the point to deal with poison replies...

Reply:
Ah, of course! - I already have `isPending(correlationId)` later in `onMessage()` - I should check it earlier 😄